h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -17,14 +17,48 @@
    *
    * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained. Only Frame and Model keys are accepted.
    */
-  public static void retain(final Key[] retainedKeys){
+  public static void retain(final Key... retainedKeys) {
     final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
     retainedSet.addAll(Arrays.asList(retainedKeys));
     // Frames and models have multiple nested keys. Those must be extracted and kept from deletion as well.
     extractNestedKeys(retainedSet);
+    final Key[] allRetainedkeys = retainedSet.toArray(new Key[retainedSet.size()]);
+
+    for (final H2ONode node : H2O.CLOUD.members()) {
+      final NodeKeysRemovalTask nodeKeysRemovalTask = new NodeKeysRemovalTask(allRetainedkeys);
+      H2O.runOnH2ONode(node, nodeKeysRemovalTask);
+    }
+  }
+
+  /**
+   * Exctracts keys a {@link Model} points to.
+   *
+   * @param retainedKeys A set of retained keys to insert the extracted {@link Model} keys to.
+   * @param model        An instance of {@link Model} to extract the keys from
+   */
+  private static void extractModelKeys(final Set<Key> retainedKeys, final Model model) {
+    Objects.requireNonNull(model);
+    if (model._parms._train != null) {
+      retainedKeys.add(model._parms._train);
+      extractFrameKeys(retainedKeys, model._parms._train.get());
+    }
+    if (model._parms._valid != null) {
+      retainedKeys.add(model._parms._valid);
+      extractFrameKeys(retainedKeys, model._parms._valid.get());
+    }
+
+    if (model._output != null && model._output.getModelMetrics() != null) {
+      for (final Key<ModelMetrics> modelMetric : model._output.getModelMetrics()) {
+        retainedKeys.add(modelMetric);
```

Comment:
Why do we need to worry about non-Frame / non-Model objects?

Reply:
This is an unfortunate leftover from the previous implementation, where this was needed, as we were iterating over all the keys and deleting whatever came our way.

I believe this can be avoided completely now. The whole `extractModelKeys` method.